Update LLVM to 21.1.2.

wffurr opened the pull request:
This required minor changes to LLVMContext construction and PGOOptions.

wffurr committed:
Update LLVM to 21.1.2.

Wraps the LLVMContext construction in LLVM 21 version check; LLVM 21 makes a
breaking change in LLVMContext construction, but WAMR still needs to
support older LLVM versions, e.g. for xtensa/esp32 support which is only
available in at most LLVM 19.

yamt commented on build-scripts/build_llvm.py:
```diff
@@ -294,17 +294,17 @@ def main():
             "repo": "https://github.com/espressif/llvm-project.git",
             "repo_ssh": "git@github.com:espressif/llvm-project.git",
-            "branch": "xtensa_release_18.1.2",
+            "branch": "xtensa_release_21.1.2",
```
no such a branch availabile

yamt commented:
> Is this good to merge? The test failures look like timeouts; one error is "fatal error: error writing to /tmp/cckAF38s.s: No space left on device" which doesn't seem relevant. Thanks!

isn't it possible llvm 21 build actually somehow requires more space and caused ENOSPC on the ci?
i guess someone needs to investigate a bit further. (sorry i have no time right now)

wffurr committed:
Free up space on Ubuntu runner for nuttx spec tests.

LLVM 21 update uses more disk space and makes the standard runner fail
with "No space left on device".  Using the [free disk space
action](https://github.com/marketplace/actions/free-disk-space-ubuntu)
to delete the unused Android, Haskell, and .NET runtimes frees up space
on the runner.

wffurr commented:
> > Is this good to merge? The test failures look like timeouts; one error is "fatal error: error writing to /tmp/cckAF38s.s: No space left on device" which doesn't seem relevant. Thanks!
> 
> isn't it possible llvm 21 build actually somehow requires more space and caused ENOSPC on the ci? i guess someone needs to investigate a bit further. (sorry i have no time right now)

I found a script to free up disk space on the GitHub Ubuntu runner.  Is that OK to use?  Seems to fix the out of space error.

lum1n0us commented:
Thanks for the contribution.
 
At the moment, we do not see a strong enough reason to upgrade WAMR’s LLVM baseline to 21, so we are not planning to change the default LLVM version in build_llvm.py for now.
 
Also, the aot_llvm.c change has already been merged through #4937.
 
As for the other two changes in this PR, they do not seem closely related to our current needs either. So unless there is some additional context or a stronger motivation for keeping this PR open, we plan to close it.
 
Thanks again for the work and for putting this together.
